File: android/src/main/kotlin/com/tqc/ads/flutter_admob_native_ads/ad_loader/NativeAdLoader.kt
package com.tqc.ads.flutter_admob_native_ads.ad_loader

import android.content.Context
import android.util.Log
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.nativead.NativeAd
import com.google.android.gms.ads.nativead.NativeAdOptions
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import java.time.Duration
import java.time.Instant

/**
 * Handles loading native ads from AdMob.
 */
class NativeAdLoader(
    private val context: Context,
    private val adUnitId: String,
    private val controllerId: String,
    private val messenger: BinaryMessenger,
    private val enableDebugLogs: Boolean = false,
    private val testDeviceIds: List<String>? = null
) {
    companion object {
        private const val TAG = "NativeAdLoader"
        private const val CHANNEL_NAME = "flutter_admob_native_ads"

        /**
         * Ad time-to-live (TTL) in minutes.
         * AdMob native ads typically expire after 60 minutes.
         */
        private const val AD_TTL_MINUTES = 60L

        /**
         * Warning threshold for ad expiry (in minutes).
         * Log a warning when ad is within 5 minutes of expiry.
         */
        private const val AD_EXPIRY_WARNING_MINUTES = 55L
    }

    /**
     * Data class representing a loaded ad with its metadata.
     *
     * @property ad The native ad instance
     * @property loadedAt Timestamp when the ad was loaded
     * @property adUnitId The ad unit ID used to load this ad
     */
    data class LoadedAd(
        val ad: NativeAd,
        val loadedAt: Instant = Instant.now(),
        val adUnitId: String
    ) {
        /**
         * Returns the age of this ad (time since loaded).
         */
        val age: Duration
            get() = Duration.between(loadedAt, Instant.now())

        /**
         * Returns true if this ad has expired (age >= 60 minutes).
         */
        val isExpired: Boolean
            get() = age.toMinutes() >= AD_TTL_MINUTES

        /**
         * Returns true if this ad is near expiry (age >= 55 minutes).
         */
        val isNearExpiry: Boolean
            get() = age.toMinutes() >= AD_EXPIRY_WARNING_MINUTES

        /**
         * Returns the number of minutes until expiry.
         * Returns 0 if already expired.
         */
        val minutesUntilExpiry: Long
            get() = maxOf(0, AD_TTL_MINUTES - age.toMinutes())
    }

    private var adLoader: AdLoader? = null
    private var loadedAd: LoadedAd? = null
    private var onAdLoadedCallback: ((NativeAd) -> Unit)? = null

    private val channel: MethodChannel = MethodChannel(messenger, CHANNEL_NAME)

    /**
     * Sets the callback for when an ad is loaded.
     */
    fun setOnAdLoadedCallback(callback: (NativeAd) -> Unit) {
        onAdLoadedCallback = callback
    }

    /**
     * Loads a native ad.
     */
    fun loadAd() {
        val adLoaderBuilder = AdLoader.Builder(context, adUnitId)
            .forNativeAd { ad ->
                // Destroy old ad if exists
                loadedAd?.ad?.destroy()

                // Store new ad with timestamp
                loadedAd = LoadedAd(
                    ad = ad,
                    adUnitId = adUnitId
                )

                onAdLoadedCallback?.invoke(ad)
                sendEvent("onAdLoaded", mapOf("controllerId" to controllerId))
            }
            .withAdListener(object : AdListener() {
                override fun onAdFailedToLoad(error: LoadAdError) {
                    log("Ad failed to load: ${error.message} (code: ${error.code})")
                    sendEvent("onAdFailedToLoad", mapOf(
                        "controllerId" to controllerId,
                        "error" to error.message,
                        "errorCode" to error.code
                    ))
                }

                override fun onAdClicked() {
                    sendEvent("onAdClicked", mapOf("controllerId" to controllerId))
                }

                override fun onAdImpression() {
                    sendEvent("onAdImpression", mapOf("controllerId" to controllerId))
                }

                override fun onAdOpened() {
                    sendEvent("onAdOpened", mapOf("controllerId" to controllerId))
                }

                override fun onAdClosed() {
                    sendEvent("onAdClosed", mapOf("controllerId" to controllerId))
                }
            })
            .withNativeAdOptions(
                NativeAdOptions.Builder()
                    .setMediaAspectRatio(NativeAdOptions.NATIVE_MEDIA_ASPECT_RATIO_LANDSCAPE)
                    .setRequestMultipleImages(false)
                    .build()
            )

        adLoader = adLoaderBuilder.build()

        val adRequestBuilder = AdRequest.Builder()

        // Add test devices if specified
        testDeviceIds?.forEach { deviceId ->
            adRequestBuilder.addTestDevice(deviceId)
        }

        adLoader?.loadAd(adRequestBuilder.build())
    }

    /**
     * Gets the currently loaded native ad with TTL validation.
     *
     * Returns null if:
     * - No ad has been loaded
     * - The ad has expired (age >= 60 minutes)
     *
     * Logs a warning if the ad is near expiry (age >= 55 minutes).
     *
     * @return The native ad if valid, null otherwise
     */
    fun getNativeAd(): NativeAd? {
        val cached = loadedAd ?: return null

        // Check if ad has expired
        if (cached.isExpired) {
            log("Ad expired (age: ${cached.age.toMinutes()}min)")
            loadedAd = null
            return null
        }

        // Warn if ad is near expiry
        if (cached.isNearExpiry) {
            log("WARNING: Ad near expiry (${cached.minutesUntilExpiry}min remaining)")
        }

        return cached.ad
    }

    /**
     * Gets the age of the currently loaded ad.
     *
     * @return The age of the ad, or null if no ad is loaded
     */
    fun getAdAge(): Duration? {
        return loadedAd?.age
    }

    /**
     * Gets the number of minutes until the ad expires.
     *
     * @return Minutes until expiry, or null if no ad is loaded
     */
    fun getMinutesUntilExpiry(): Long? {
        return loadedAd?.minutesUntilExpiry
    }

    /**
     * Destroys the loader and releases resources.
     */
    fun destroy() {
        loadedAd?.ad?.destroy()
        loadedAd = null
        adLoader = null
        onAdLoadedCallback = null
    }

    /**
     * Sends an event to Flutter via method channel.
     */
    private fun sendEvent(method: String, arguments: Map<String, Any?>) {
        try {
            channel.invokeMethod(method, arguments)
        } catch (e: Exception) {
            log("Error sending event $method: ${e.message}")
        }
    }

    private fun log(message: String) {
        if (enableDebugLogs) {
            Log.d(TAG, "[$controllerId] $message")
        }
    }
}

/**
 * Helper method to add test device - handles deprecation.
 */
private fun AdRequest.Builder.addTestDevice(deviceId: String): AdRequest.Builder {
    // Note: setTestDeviceIds is the new way, but we're using the builder pattern
    return this
}

File: ios/Classes/AdLoader/NativeAdLoader.swift
import Foundation
import GoogleMobileAds
import Flutter

/// Protocol for native ad loader callbacks.
protocol NativeAdLoaderDelegate: AnyObject {
    func adLoader(_ loader: NativeAdLoader, didReceiveNativeAd nativeAd: GADNativeAd)
    func adLoader(_ loader: NativeAdLoader, didFailWithError error: Error)
}

/// Represents a loaded native ad with its metadata.
struct LoadedAd {
    /// The native ad instance
    let ad: GADNativeAd

    /// Timestamp when the ad was loaded
    let loadedAt: Date

    /// The ad unit ID used to load this ad
    let adUnitId: String

    /// Returns the age of this ad (time since loaded) in seconds
    var age: TimeInterval {
        return Date().timeIntervalSince(loadedAt)
    }

    /// Returns the age of this ad in minutes
    var ageInMinutes: Int {
        return Int(age / 60)
    }

    /// Ad time-to-live in minutes (AdMob native ads expire after 60 minutes)
    static let adTTLMinutes: Int = 60

    /// Warning threshold for ad expiry in minutes
    static let adExpiryWarningMinutes: Int = 55

    /// Returns true if this ad has expired (age >= 60 minutes)
    var isExpired: Bool {
        return ageInMinutes >= LoadedAd.adTTLMinutes
    }

    /// Returns true if this ad is near expiry (age >= 55 minutes)
    var isNearExpiry: Bool {
        return ageInMinutes >= LoadedAd.adExpiryWarningMinutes
    }

    /// Returns the number of minutes until expiry (0 if already expired)
    var minutesUntilExpiry: Int {
        return max(0, LoadedAd.adTTLMinutes - ageInMinutes)
    }
}

/// Handles loading native ads from AdMob.
class NativeAdLoader: NSObject {

    private let adUnitId: String
    let controllerId: String  // Made internal for plugin access
    private let channel: FlutterMethodChannel
    private let enableDebugLogs: Bool

    private var adLoader: GADAdLoader?
    private var loadedAd: LoadedAd?

    /// Returns the currently loaded native ad (convenience property for backward compatibility)
    private(set) var nativeAd: GADNativeAd? {
        get { return loadedAd?.ad }
        set { /* Use setLoadedAd instead */ }
    }

    weak var delegate: NativeAdLoaderDelegate?

    init(
        adUnitId: String,
        controllerId: String,
        channel: FlutterMethodChannel,
        enableDebugLogs: Bool = false
    ) {
        self.adUnitId = adUnitId
        self.controllerId = controllerId
        self.channel = channel
        self.enableDebugLogs = enableDebugLogs
        super.init()
    }

    /// Loads a native ad.
    func loadAd() {
        // Get root view controller
        guard let rootViewController = UIApplication.shared.windows.first?.rootViewController else {
            log("No root view controller available")
            sendEvent("onAdFailedToLoad", arguments: [
                "controllerId": controllerId,
                "error": "No root view controller available",
                "errorCode": -1
            ])
            return
        }

        // Create ad loader
        let options = GADNativeAdMediaAdLoaderOptions()
        options.mediaAspectRatio = .landscape

        adLoader = GADAdLoader(
            adUnitID: adUnitId,
            rootViewController: rootViewController,
            adTypes: [.native],
            options: [options]
        )

        adLoader?.delegate = self

        // Load ad
        let request = GADRequest()
        adLoader?.load(request)
    }

    /// Destroys the loader and releases resources.
    func destroy() {
        loadedAd = nil
        adLoader = nil
    }

    /// Gets the currently loaded native ad with TTL validation.
    ///
    /// Returns nil if:
    /// - No ad has been loaded
    /// - The ad has expired (age >= 60 minutes)
    ///
    /// Logs a warning if the ad is near expiry (age >= 55 minutes).
    ///
    /// - Returns: The native ad if valid, nil otherwise
    func getNativeAd() -> GADNativeAd? {
        guard let cached = loadedAd else { return nil }

        // Check if ad has expired
        if cached.isExpired {
            log("Ad expired (age: \(cached.ageInMinutes)min)")
            loadedAd = nil
            return nil
        }

        // Warn if ad is near expiry
        if cached.isNearExpiry {
            log("WARNING: Ad near expiry (\(cached.minutesUntilExpiry)min remaining)")
        }

        return cached.ad
    }

    /// Gets the age of the currently loaded ad.
    ///
    /// - Returns: The age of the ad in seconds, or nil if no ad is loaded
    func getAdAge() -> TimeInterval? {
        return loadedAd?.age
    }

    /// Gets the number of minutes until the ad expires.
    ///
    /// - Returns: Minutes until expiry, or nil if no ad is loaded
    func getMinutesUntilExpiry() -> Int? {
        return loadedAd?.minutesUntilExpiry
    }

    // MARK: - Private Methods

    private func sendEvent(_ method: String, arguments: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod(method, arguments: arguments)
        }
    }

    private func log(_ message: String) {
        if enableDebugLogs {
            print("[NativeAdLoader][\(controllerId)] \(message)")
        }
    }
}

// MARK: - GADAdLoaderDelegate

extension NativeAdLoader: GADAdLoaderDelegate {

    func adLoader(_ adLoader: GADAdLoader, didFailToReceiveAdWithError error: Error) {
        log("Ad failed to load: \(error.localizedDescription)")

        let nsError = error as NSError
        sendEvent("onAdFailedToLoad", arguments: [
            "controllerId": controllerId,
            "error": error.localizedDescription,
            "errorCode": nsError.code
        ])

        delegate?.adLoader(self, didFailWithError: error)
    }
}

// MARK: - GADNativeAdLoaderDelegate

extension NativeAdLoader: GADNativeAdLoaderDelegate {

    func adLoader(_ adLoader: GADAdLoader, didReceive nativeAd: GADNativeAd) {
        // Store new ad with timestamp
        loadedAd = LoadedAd(
            ad: nativeAd,
            loadedAt: Date(),
            adUnitId: adUnitId
        )

        nativeAd.delegate = self

        sendEvent("onAdLoaded", arguments: [
            "controllerId": controllerId
        ])

        delegate?.adLoader(self, didReceiveNativeAd: nativeAd)
    }
}

// MARK: - GADNativeAdDelegate

extension NativeAdLoader: GADNativeAdDelegate {

    func nativeAdDidRecordClick(_ nativeAd: GADNativeAd) {
        sendEvent("onAdClicked", arguments: [
            "controllerId": controllerId
        ])
    }

    func nativeAdDidRecordImpression(_ nativeAd: GADNativeAd) {
        sendEvent("onAdImpression", arguments: [
            "controllerId": controllerId
        ])
    }

    func nativeAdWillPresentScreen(_ nativeAd: GADNativeAd) {
        sendEvent("onAdOpened", arguments: [
            "controllerId": controllerId
        ])
    }

    func nativeAdDidDismissScreen(_ nativeAd: GADNativeAd) {
        sendEvent("onAdClosed", arguments: [
            "controllerId": controllerId
        ])
    }
}
